### PR TITLE
Cosmo strict copy

### PR DESCRIPTION
**Pull request by nstarman**
### Description

The storing of initialization arguments was a good idea for making cloning easier but a few things have changed / are changing.
1. The Parameters object has been added, which regularizes how most inputs to a cosmology are handled.
2. I plan to allow a Cosmology to be array-valued.
    For large arrays, e.g. an MCMC chain, storing the initialization arguments doubles the memory footprint.

This new method changes `_init_arguments` from a dictionary to a `property`. Subclasses can override the property if needed. Parameters have a method ``inv_validate`` to reverse ``validate`` and get the original value.

I haven't yet decided how fully featured the ``inv_validate`` should be: a registry system like ``fvalidate`` or the current minimal implementation. The only think a registry system for ``inv_validate`` would change is that ``m_nu`` wouldn't need to be special-cased in ``FLRW._init_arguments``.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [ ] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

File: astropy/cosmology/tests/test_parameter.py [hunk 1]
```diff
@@ -400,6 +400,14 @@ def func(cosmology, param, value):
         finally:
             param.__class__._registry_validators.pop("newvalidator", None)
 
+    def test_Parameter_inv_validate(self, cosmo, param):
+        """Test :meth:`astropy.cosmology.Parameter.inv_validate`."""
+        # inverse validate to get original value
+        init_arg = param.inv_validate(cosmo)
+        expected = {param.name: param.__get__(cosmo)}
+        assert init_arg == expected
+        assert init_arg["param"] == 15 * u.m
```
Comment (nstarman) [inline astropy/cosmology/tests/test_parameter.py:409]:
```suggestion
        expected = {"param": 15 * u.m}
        assert init_arg == expected
```

**Review by mhvk — COMMENTED** [review]
I think this is a good idea - the storing and introspection of those arguments always felt rather odd to me. I wonder, though, whether `inv_validate` is actually needed. It feels like a large complication, also for adding new parameters, for very little gain. Why not just use the validated values themselves? As long as those give the same cosmology when used to initialize a new one, I do not really see the problem.

**Comment by nstarman** [comment]
It's largely because the ``_init_arguments`` decoupled instantiation from the processed inputs. Previously the ``property`` (now `Parameter`) on Cosmology had to match the input arguments.

```python
class A()
    def __init__(self, arg1, arg2):
        ...
    @property
    def arg1(self): ...
    @property
    def arg2(self): ...
```

With ``_init_arguments``, that's no longer the case. The following could be cloned:

```python
class A()
    def __init__(self, turnsintoarg1, arg2):
        ...
    arg1 = Parameter()
    arg2 = Parameter()
```

Getting rid of ``_init_arguments`` means there needs to be some way to remap the ``arg1`` Parameter to ``turnsintoarg1``. Hence ``inv_validate``.

**Comment by mhvk** [comment]
I would go with your "OTOH" here and just assume nobody has that kind of cosmology, or, if they have, they do not rely on `_init_arguments`... Indeed, perhaps an idea is to deprecate it in this PR?

**Comment by nstarman** [comment]
> I would go with your "OTOH" here and just assume nobody has that kind of cosmology, or, if they have, they do not rely on `_init_arguments`...

SGTM. I also doubt it will happen. I don't think the possibility was well advertised when I refactored ``clone``.

> Indeed, perhaps an idea is to deprecate it in this PR?

``_init_arguments`` is used in both ``clone`` and ``to_mapping``, so I think it's worth keeping around as a private attribute.

**Comment by nstarman** [comment]
Ok. I've rolled back the ``inv_validate``. If we ever want to support more complex init signatures ``inv_validate`` can be reintroduced.

File: astropy/cosmology/io/tests/test_ecsv.py [hunk 2]
```diff
@@ -169,7 +169,7 @@ def test_fromformat_ecsv_subclass_partial_info(self, cosmo_cls, cosmo, read, wri
         # not equal, because Tcmb0 is changed
         assert got != cosmo
         assert got.Tcmb0 == cosmo_cls._init_signature.parameters["Tcmb0"].default
-        assert got.clone(name=cosmo.name, Tcmb0=cosmo.Tcmb0) == cosmo
+        assert got.clone(name=cosmo.name, Tcmb0=cosmo.Tcmb0, m_nu=cosmo.m_nu) == cosmo
```
Comment (mhvk) [inline astropy/cosmology/io/tests/test_ecsv.py:172]: Why is this necessary? You didn't change `m_nu`. Maybe related to the override of `_init_arguments` above?

Note that this is not a big deal, but if there is a reason, add it to the comment on line 169.

(obviously, the same comment for all the other instances...)

**Review by mhvk — COMMENTED** [review]
This now looks like a very nice clean-up. I only do not understand why `m_nu` is treated differently.

File: astropy/cosmology/flrw.py [hunk 3]
```diff
@@ -268,6 +268,13 @@ def __init__(self, H0, Om0, Ode0, Tcmb0=0.0*u.K, Neff=3.04, m_nu=0.0*u.eV,
         self._inv_efunc_scalar = self.inv_efunc
         self._inv_efunc_scalar_args = ()
 
+    @property
+    def _init_arguments(self):
+        ps = super()._init_arguments
+        if ps["m_nu"] is None:
```
[the remark this answers is not in the document]
Comment (nstarman) [reply astropy/cosmology/flrw.py]: Ahh. This is from a followup PR. I didn't separate the two properly. Thanks for the catch. Yes, it will give different results for ``cosmo.to_format("astropy.model", method=...)``, but that's for a followup PR.

[the remark this answers is not in the document]
Comment (nstarman) [reply astropy/cosmology/flrw.py]: Expect to see this exact diff soon :)

[the remark this answers is not in the document]
Comment (nstarman) [reply astropy/cosmology/flrw.py]: But I'm removing it from here.

File: astropy/cosmology/io/tests/test_ecsv.py (hunk 2 above)
Comment (mhvk) [reply astropy/cosmology/io/tests/test_ecsv.py:172]: OK, I see now!

**Review by mhvk — APPROVED** [review]
OK, looks all fine now!  Approving and merging...
